common/src/main/java/io/netty/util/internal/PlatformDependent0.java

```diff
@@ -415,7 +437,605 @@ static void freeMemory(long address) {
+                    } catch (SecurityException e) {
+                        logger.info("No permissions to get StringBuilder value array." +
+                                " Hash code optimizations for this type are disabled.", e);
+                    }
+                    return null;
+                }
+            });
+            STRING_VALUE_FIELD_OFFSET = stringValueField == null ?
+                    -1 : PlatformDependent0.objectFieldOffset(stringValueField);
+            STRING_BUILDER_VALUE_FIELD_OFFSET = stringBuilderValueField == null ?
+                    -1 : PlatformDependent0.objectFieldOffset(stringBuilderValueField);
+        }
+
+        @Override
+        public final int emptyHashValue() {
+            return EMPTY_HASH;
+        }
+
+        @Override
+        public final int hashCodeAsBytes(CharSequence data, int startPos, int endPos) {
+            // CharSequence by itself does not provide access to any underlying data structure, and using
+            // charAt can be costly. For the common types in the CharSequence we can provide an optimized
+            // hashCode by direclty using the underlying char[].
+            if (data.getClass() == String.class && STRING_VALUE_FIELD_OFFSET != -1) {
+                return hashCodeAsBytes(
+                        (char[]) UNSAFE.getObject(data, STRING_VALUE_FIELD_OFFSET), startPos, endPos);
+            }
+            if (data.getClass() == StringBuilder.class && STRING_BUILDER_VALUE_FIELD_OFFSET != -1) {
+                return hashCodeAsBytes(
+                        (char[]) UNSAFE.getObject(data, STRING_BUILDER_VALUE_FIELD_OFFSET), startPos, endPos);
+            }
+            return hashCodeAsBytesDirect(data, startPos, endPos);
+        }
+
+        protected abstract int hashCodeAsBytesDirect(CharSequence data, int startPos, int endPos);
+    }
+
+    private static final class NettyHashAscii extends AbstractNettyHashAscii {
+        public static final NettyHashAscii INSTANCE = new NettyHashAscii();
+        private NettyHashAscii() { }
+        @Override
+        public int hashCode(byte[] bytes, int startPos, int endPos) {
+            final int len = endPos - startPos;
+            final int remainingBytes = len & 7;
+            final long baseOffset = startPos + BYTE_ARRAY_BASE_OFFSET;
+            final long rend = remainingBytes + baseOffset;
+            long hash = HASH_OFFSET;
+            for (long i = len - 8 + baseOffset; i >= rend; i -= 8) {
+                hash ^= UNSAFE.getLong(bytes, i);
+                hash *= HASH_PRIME;
+            }
+            // Account for each case of remaining bytes independently. Conditional statements and decrementing a
+            // variable to share code proved to generate measurable amount of overhead for small inputs which are
+            // expected to be common (i.e. Header Names).
+            switch (remainingBytes) {
+                case 7: {
```

Comment:
@louiscryan @nmittler - Thought we didn't indent :).

@normanmaurer - I think HTTP/2 currently indents, but I will fix this here (I prefer not indenting).